### PR TITLE
Make valgrind suppressions great again

### PR DESCRIPTION
**Pull request by petermax2**
# Purpose

Clean up the `valgrind.suppression` file by replacing concrete patterns with more generic ones.
This change aims to exclude all calls to crypto libraries from memory checking.

See #932 for full discussion.

# Checklist

Please only check relevant points.
For docu fixes, spell checking and similar nothing
needs to be checked.

- [x] commit messages are fine (with references to issues)
- [x] I ran all tests and everything went fine
- [ ] I added unit tests
- [ ] affected documentation is fixed
- [ ] I added code comments, logging, and assertions
- [ ] meta data is updated (e.g. README.md of plugins)

@markus2330 please review my pull request

### CODE REVIEW

File: tests/valgrind.suppression [hunk 1]
```diff
@@ -1,131 +1,87 @@
...
-   fun:_ZN5Botan13Library_State10initializeEb
-   fun:_ZN5Botan18LibraryInitializer10initializeERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE
-   obj:*
-   obj:*
-   fun:elektraPluginOpen
-   fun:test_crypto_operations
-   fun:main
+   match-leak-kinds: all
+   ...
+   obj:*libbotan*
```
Comment (markus2330) [inline tests/valgrind.suppression]: Yes, that looks like it should be ;)

File: tests/valgrind.suppression [hunk 2]
```diff
@@ -1,131 +1,87 @@
...
-   fun:elektraPluginOpen
 }
 {
    <insert_a_suppression_name_here>
    Memcheck:Cond
-   fun:is_overlap
-   fun:memcpy@@GLIBC_2.14
-   fun:keySetRaw
+   ...
    fun:elektraCryptoGcryDecrypt
```
Comment (markus2330) [inline tests/valgrind.suppression]: Is this unfixable? Are you sure its not a mistake of CryptoGcryDecrypt or keySetRaw?

File: tests/valgrind.suppression [hunk 3]
```diff
@@ -1,131 +1,87 @@
...
-   fun:_ZN5Botan13Library_State10initializeEb
-   fun:_ZN5Botan18LibraryInitializer10initializeERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE
-   obj:*
-   obj:*
-   fun:elektraPluginOpen
-   fun:test_crypto_operations
-   fun:main
+   match-leak-kinds: all
+   ...
+   fun:CRYPTO_*
```
Comment (markus2330) [inline tests/valgrind.suppression]: Can you limit to `obj:*crypto*`? (`obj` might be useful at some other places, too?)

**Comment by markus2330** [comment]
jenkins build incremental please

**Comment by markus2330** [comment]
jenkins build gcc please

**Comment by markus2330** [comment]
jenkins build gcc-configure-debian please

**Comment by markus2330** [comment]
jenkins build gcc-configure-debian-shared please

**Comment by markus2330** [comment]
jenkins build gcc-configure-debian-wheezy please

**Comment by markus2330** [comment]
jenkins build gcc-configure-debian-stretch please

File: tests/valgrind.suppression (hunk 2 above)
Comment (petermax2) [inline tests/valgrind.suppression]: I tried several things in `CryptoGcryDecrypt` and came to the conclusion that `gcry_cipher_decrypt` is causing the conditional jump. The valgrind message comes in different variations, including one which has no `keySetRaw` as a frame.

File: tests/valgrind.suppression (hunk 3 above)
Comment (petermax2) [inline tests/valgrind.suppression]: It does not cause problems on Debian stable. Let's see how other build variants behave. Adapted in a87d653fddd7b7853b059358bdc2a5b2b612186e .

**Comment by petermax2** [comment]
@markus2330 3 tests failed for my previous commit. But it seems like gpg / gpg2 is not available. Has the `gnupg2`package been installed to all Jenkins instances?

**Comment by petermax2** [comment]
Is it possible to receive the build artifacts (`DynamicAnalysis.xml`) for the build variant `elektra-gcc`?

File: tests/valgrind.suppression (hunk 2 above)
Comment (markus2330) [inline tests/valgrind.suppression]: Ok, then its ok.

**Comment by markus2330** [comment]
You can always download every file from the workspace.

**Comment by petermax2** [comment]
> You can always download every file from the workspace.

There is no link to the Workplace directory for this build variant. It looks different to stable and unstable, etc.

**Comment by markus2330** [comment]
On which agents are the packages missing?

**Comment by markus2330** [comment]
http://community.markus-raab.org:8080/job/elektra-gcc/ws/

**Comment by petermax2** [comment]
> On which agents are the packages missing?

debian-wheezy-mr

> http://community.markus-raab.org:8080/job/elektra-gcc/ws/

Thanks!

**Comment by markus2330** [comment]
There is a package conflict on debian-wheezy-mr with gnupg2 and libgcrypt11, I added the task to #160.

**Comment by petermax2** [comment]
> debian-wheezy-mr

debian-strech-mr is also affected.

**Comment by petermax2** [comment]
jenkins build elektra-gcc please

**Comment by petermax2** [comment]
jenkins build gcc please

**Comment by petermax2** [comment]
173ced48caf34220f27c175dc9e96394b60cf0a2 successfully suppresses the libstdc++ specific valgrind error (the problem originates in a Botan constructor, so it should be OK to suppress it).

@markus2330 I am ready to merge. Do you want to wait until the build agents are fixed?

**Comment by markus2330** [comment]
jenkins build please

**Comment by markus2330** [comment]
jenkins build stable please

**Comment by markus2330** [comment]
jenkins build stretch please

**Comment by markus2330** [comment]
jenkins build gcc please

**Comment by markus2330** [comment]
Why is it necessary that gnupg2 (and not only gnupg) is installed?

I installed it now, can you check if everything is okay now?

**Comment by petermax2** [comment]
> Why is it necessary that gnupg2 (and not only gnupg) is installed?

The PATH seems to be empty in the build jobs, so the gpg module can not find any executables. `/usr/bin/gpg2` is used as kind of a last-resort attempt to get the test cases running.

**Comment by petermax2** [comment]
jenkins build stretch please

**Comment by markus2330** [comment]
Why not trying `/usr/bin/gpg`, too? Did you look at the build jobs? Is everything fine now?

**Comment by petermax2** [comment]
> Why not trying /usr/bin/gpg, too?

will open an issue for that.

> Did you look at the build jobs? Is everything fine now?

`elektra-gcc` is fine now, but `stretch` and `wheezy` did not start.

**Comment by markus2330** [comment]
jenkins build stretch please

**Comment by markus2330** [comment]
jenkins build gcc-configure-debian-wheezy please

**Comment by markus2330** [comment]
jenkins build gcc-configure-debian-stretch please

**Comment by petermax2** [comment]
`stretch` reports problems, that have no connection to my crypto-plugin but rather with mathcheck:

	==23580== Conditional jump or move depends on uninitialised value(s)
	==23580==    at 0x56AAF63: __printf_fp_l (printf_fp.c:1212)
	==23580==    by 0x56A87D8: vfprintf (vfprintf.c:1631)
	==23580==    by 0x56CE0B8: vsnprintf (vsnprintf.c:114)
	==23580==    by 0x56AEA81: snprintf (snprintf.c:33)
	==23580==    by 0x5E03180: elektraFtoA (floathelper.c:24)
	==23580==    by 0x5E02B36: elektraMathcheckSet (mathcheck.c:369)
	==23580==    by 0x402CB5: main (testmod_mathcheck.c:73)
	==23580== 

and the second one:

	==23580==    at 0x56AB025: round_away (rounding-mode.h:52)
	==23580==    by 0x56AB025: __printf_fp_l (printf_fp.c:979)
	==23580==    by 0x56A87D8: vfprintf (vfprintf.c:1631)
	==23580==    by 0x56CE0B8: vsnprintf (vsnprintf.c:114)
	==23580==    by 0x56AEA81: snprintf (snprintf.c:33)
	==23580==    by 0x5E03180: elektraFtoA (floathelper.c:24)
	==23580==    by 0x5E02B36: elektraMathcheckSet (mathcheck.c:369)
	==23580==    by 0x402CB5: main (testmod_mathcheck.c:73)

Should I suppress it?

The crypto-plugin seems to be OK now.

**Comment by markus2330** [comment]
@petermax2 not as part of this PR and first we have to make sure its not a problem in the code. Thanks for the restructuring of the valgrind suppression, you really made them great again ;)

@tom-wa please fix the memory problem in mathcheck

**Comment by markus2330** [comment]
@tom-wa memory problem is described in #942, I will unassign you here.

**Pull request merged**